File: macos/Sources/Features/Setup/ClaudeHookSpec.swift
// macos/Sources/Features/Setup/ClaudeHookSpec.swift
import Foundation

/// Claude stores hooks in the SHARED ~/.claude/settings.json (no auto-loaded
/// hooks dir), so Ghoztty merges a fragment under the `hooks` key and tracks
/// ownership by the banner-script invocation signature.
struct ClaudeHookSpec: HookSpec {
    let ownership: HookOwnership = .mergedFragment
    let marker = GhosttyManagedMarker.token

    func hookFileURL(homeDirectoryURL: URL) -> URL {
        homeDirectoryURL.appendingPathComponent(".claude/settings.json")
    }

    // renderedFile is unused for mergedFragment; the component uses the merge
    // helpers instead. Provide the whole-file form for protocol conformance/tests.
    func renderedFile(bannerScriptPath: String) -> String {
        let json = Self.merge(into: [:], bannerScriptPath: bannerScriptPath)
        let data = (try? JSONSerialization.data(withJSONObject: json, options: [.prettyPrinted, .sortedKeys])) ?? Data()
        return String(data: data, encoding: .utf8) ?? ""
    }

    /// Ownership signature: any hook command that invokes our banner script.
    static func signature(_ bannerScriptPath: String) -> String { bannerScriptPath }

    /// Ghoztty's contribution to the shared `hooks` map: one element per event.
    /// The map is keyed by event name; each value is the array of hook elements
    /// Ghoztty appends for that event. Each command carries a `timeout` (seconds)
    /// so a hook — notably Stop's best-effort `gh pr view` — can never stall the
    /// agent's turn; Claude reads the `timeout` key (Copilot uses `timeoutSec`).
    static func hooksBlock(bannerScriptPath: String) -> [String: [[String: Any]]] {
        func command(_ purpose: HookPurpose) -> [String: Any] {
            ["type": "command",
             "command": HookCommand.perEvent(purpose: purpose, bannerScriptPath: bannerScriptPath, runtime: .claude),
             "timeout": 10]
        }
        func entry(_ purpose: HookPurpose) -> [[String: Any]] {
            [["hooks": [command(purpose)]]]
        }
        return [
            "SessionStart": [["matcher": "startup|clear", "hooks": [command(.sessionStart)]]],
            "UserPromptSubmit": entry(.promptSubmit),
            "Stop": entry(.stop),
        ]
    }

    /// Serialize a single hook element deterministically for signature matching
    /// and set comparison (sorted keys, literal slashes so the script path is
    /// byte-comparable to `signature`).
    private static func serialized(_ element: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(
                  withJSONObject: element, options: [.sortedKeys, .withoutEscapingSlashes]),
              let text = String(data: data, encoding: .utf8) else { return "" }
        return text
    }

    /// Merge Ghoztty's hooks into the user's shared settings WITHOUT clobbering
    /// their own hooks. Only Ghoztty's three events are touched, and within each
    /// only prior Ghoztty elements (matched by the script-path signature) are
    /// replaced — every other event and every non-Ghoztty element is preserved.
    static func merge(into json: [String: Any], bannerScriptPath: String) -> [String: Any] {
        var out = json
        var hooks = (out["hooks"] as? [String: Any]) ?? [:]
        let sig = signature(bannerScriptPath)
        for (event, ghosttyElements) in hooksBlock(bannerScriptPath: bannerScriptPath) {
            var arr = (hooks[event] as? [[String: Any]]) ?? []
            arr.removeAll { serialized($0).contains(sig) }
            arr.append(contentsOf: ghosttyElements)
            hooks[event] = arr
        }
        out["hooks"] = hooks
        return out
    }

    /// Remove ONLY Ghoztty's own elements (matched by signature) from each event,
    /// pruning an event key that becomes empty and the whole `hooks` map if it
    /// becomes empty. The user's other events and elements are left intact.
    static func removeFragment(from json: [String: Any], bannerScriptPath: String) -> [String: Any] {
        var out = json
        guard var hooks = out["hooks"] as? [String: Any] else { return out }
        let sig = signature(bannerScriptPath)
        for (event, value) in hooks {
            guard var arr = value as? [[String: Any]] else { continue }
            arr.removeAll { serialized($0).contains(sig) }
            if arr.isEmpty {
                hooks.removeValue(forKey: event)
            } else {
                hooks[event] = arr
            }
        }
        if hooks.isEmpty {
            out.removeValue(forKey: "hooks")
        } else {
            out["hooks"] = hooks
        }
        return out
    }

    /// State of ONLY Ghoztty's fragment: collect the Ghoztty elements (by
    /// signature) from the three events and compare them against what we would
    /// install. The user's other hooks vary and are never part of the compare.
    static func fragmentState(in json: [String: Any], bannerScriptPath: String) -> ComponentInstallState {
        let sig = signature(bannerScriptPath)
        let hooks = (json["hooks"] as? [String: Any]) ?? [:]
        var found = false
        var matches = true
        for (event, wantElements) in hooksBlock(bannerScriptPath: bannerScriptPath) {
            let existing = (hooks[event] as? [[String: Any]]) ?? []
            let ours = existing.filter { serialized($0).contains(sig) }
            if !ours.isEmpty { found = true }
            let wantSet = Set(wantElements.map { serialized($0) })
            let haveSet = Set(ours.map { serialized($0) })
            if wantSet != haveSet { matches = false }
        }
        guard found else { return .notInstalled }
        return matches ? .installed : .outdated
    }

    /// The plugin's name in Claude's marketplace. The marketplace half of a
    /// manifest key is deliberately NOT part of the match: the same plugin is
    /// registered through more than one marketplace in practice
    /// (`ghoztty@dzearing-claude-marketplace` and `ghoztty@ghoztty-claude-plugin`
    /// side by side), and either one owns the runtime.
    static let externalPluginName = "ghoztty"

    /// Is Claude's own `ghoztty` plugin installed — i.e. does something other
    /// than Ghoztty already own this runtime's skills and hooks?
    ///
    /// Parsed, never substring-matched. The manifest records each install's
    /// `installPath` and (at project scope) its `projectPath`, so a bare
    /// `contains("ghoztty")` reports true for ANY plugin the user installed
    /// while sitting in a ghoztty checkout — and Ghoztty would then silently
    /// decline to install, believing its own plugin was already there.
    func isExternalPluginInstalled(homeDirectoryURL: URL, fileManager: FileManager) -> Bool {
        !Self.externalPluginRegistrations(homeDirectoryURL: homeDirectoryURL, fileManager: fileManager).isEmpty
    }

    /// Every manifest key registering the external plugin, e.g.
    /// `ghoztty@dzearing-claude-marketplace`. Plural because the same plugin is
    /// registered through more than one marketplace in practice, and each
    /// registration is separately installed — and separately uninstallable.
    ///
    /// One parse site: detection asks whether this is empty, migration asks what
    /// is in it.
    static func externalPluginRegistrations(homeDirectoryURL: URL, fileManager: FileManager) -> [String] {
        let url = homeDirectoryURL.appendingPathComponent(".claude/plugins/installed_plugins.json")
        guard let data = fileManager.contents(atPath: url.path),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [] }

        // `"version": 2` — `plugins` is an object keyed by `<name>@<marketplace>`.
        if let byKey = json["plugins"] as? [String: Any] {
            return byKey.keys.filter { name(ofRegistration: $0) == externalPluginName }
        }
        // Older shape — `plugins` is an array of entries carrying `name`.
        if let list = json["plugins"] as? [[String: Any]] {
            return list.compactMap { $0["name"] as? String }.filter { $0 == externalPluginName }
        }
        // A shape we do not recognize: report absent rather than guessing, so a
        // manifest revision cannot silently suppress the integration. The
        // coexistence cost of a false negative is a duplicate skill; the cost of
        // a false positive is Ghoztty refusing to install at all.
        return []
    }

    private static func name(ofRegistration key: String) -> String {
        key.split(separator: "@").first.map(String.init) ?? key
    }
}

File: src/apprt/win32/ClaudeHookSpec.zig
//! Ghoztty's hook fragment for Claude Code (T868, the win32 half of Mac's
//! `ClaudeHookSpec.swift`). Claude stores hooks in the SHARED
//! `~/.claude/settings.json` (no auto-loaded hooks dir), so Ghoztty merges a
//! fragment under the `hooks` key and tracks ownership by the banner-script
//! invocation signature — never a whole-file marker, because the file is the
//! user's.
//!
//! Ownership signature: any hook element whose deterministic serialization
//! contains our owned hooks DIRECTORY. Deliberately the directory rather
//! than the banner script path — Ghoztty installs a second script there
//! (`ghoztty-activity-state.sh`), and a banner-path signature would leave
//! every activity-state element unrecognized, so uninstall would strand them
//! and `fragmentState` would never report installed. An older banner-only
//! install still matches, since its path is inside this directory.
//!
//! External-plugin detection (`installed_plugins.json`) is deliberately NOT
//! here: it belongs to the registry/factory slice (T869), which is where
//! Mac's factory consumes it.
//!
//! All JSON is `std.json` over `std.json.Value` with the deterministic
//! serializer in `stable_json.zig`; functions expect an ARENA allocator (the
//! parsed/merged values freely alias each other and the input). No OS
//! imports, so the unit tests run in every app-runtime lane.
const std = @import("std");
const Allocator = std.mem.Allocator;

const hook_scripts = @import("hook_scripts.zig");
const hook_spec = @import("hook_spec.zig");
const stable_json = @import("stable_json.zig");
const marker_mod = @import("managed_marker.zig");

pub const InstallState = marker_mod.InstallState;

/// Home-relative path of Claude's shared settings file (forward slashes:
/// `std.fs` accepts them on Windows and the spelling stays platform-neutral).
pub const settings_sub_path = ".claude/settings.json";

/// The ownership signature for a given banner-script path (see module doc).
pub fn signature(banner_script_path: []const u8) []const u8 {
    return hook_scripts.directory(banner_script_path);
}

/// Ghoztty's contribution to the shared `hooks` map: event name → the array
/// of hook elements Ghoztty appends for that event. Each command carries a
/// `timeout` (seconds) so a hook can never stall the agent's turn (Claude
/// reads `timeout`; Copilot uses `timeoutSec`).
pub fn hooksBlock(
    arena: Allocator,
    banner_script_path: []const u8,
) Allocator.Error!std.json.ObjectMap {
    var block = std.json.ObjectMap.init(arena);

    // Banner: the sticky per-pane overlay.
    // Activity: the idle/busy/needs_input state machine. Both run on
    // SessionStart and Stop, in that order, as separate elements — the
    // banner's own matcher is `startup|clear` (a resume keeps its banner),
    // while the activity sweep must run on EVERY session start so a killed
    // session's leaked agent markers are always reaped.
    try block.put("SessionStart", try eventArray(arena, &.{
        .{ .matcher = "startup|clear", .purpose = .session_start },
        .{ .purpose = .activity_session_start },
    }, banner_script_path));
    try block.put("UserPromptSubmit", try eventArray(arena, &.{
        .{ .purpose = .prompt_submit },
    }, banner_script_path));
    // Anything that blocks on the user outranks everything else.
    try block.put("PreToolUse", try eventArray(arena, &.{
        .{ .matcher = "AskUserQuestion|ExitPlanMode", .purpose = .activity_pause },
    }, banner_script_path));
    try block.put("Notification", try eventArray(arena, &.{
        .{ .purpose = .activity_pause },
    }, banner_script_path));
    try block.put("PermissionRequest", try eventArray(arena, &.{
        .{ .purpose = .activity_pause },
    }, banner_script_path));
    // Catch-all: on the main thread this undoes a pause; inside a subagent
    // it only beats that agent's liveness marker.
    try block.put("PostToolUse", try eventArray(arena, &.{
        .{ .purpose = .activity_tool_tick },
    }, banner_script_path));
    try block.put("SubagentStart", try eventArray(arena, &.{
        .{ .purpose = .activity_agent_start },
    }, banner_script_path));
    try block.put("SubagentStop", try eventArray(arena, &.{
        .{ .purpose = .activity_agent_stop },
    }, banner_script_path));
    try block.put("Stop", try eventArray(arena, &.{
        .{ .purpose = .stop },
        .{ .purpose = .activity_settle },
    }, banner_script_path));

    return block;
}

const ElementSpec = struct {
    matcher: ?[]const u8 = null,
    purpose: hook_spec.HookPurpose,
};

/// `[{"matcher": …, "hooks": [{"type":"command", …}]}, …]` for one event.
fn eventArray(
    arena: Allocator,
    specs: []const ElementSpec,
    banner_script_path: []const u8,
) Allocator.Error!std.json.Value {
    var arr = std.json.Array.init(arena);
    for (specs) |spec| {
        var command = std.json.ObjectMap.init(arena);
        try command.put("type", .{ .string = "command" });
        try command.put("command", .{
            .string = try hook_spec.perEventCommand(arena, spec.purpose, banner_script_path, .claude),
        });
        try command.put("timeout", .{ .integer = 10 });

        var hooks_arr = std.json.Array.init(arena);
        try hooks_arr.append(.{ .object = command });

        var element = std.json.ObjectMap.init(arena);
        if (spec.matcher) |m| try element.put("matcher", .{ .string = m });
        try element.put("hooks", .{ .array = hooks_arr });
        try arr.append(.{ .object = element });
    }
    return .{ .array = arr };
}

/// Whether a hook element is one of ours: its deterministic serialization
/// contains the hooks-directory signature.
fn isOurs(arena: Allocator, element: std.json.Value, sig: []const u8) Allocator.Error!bool {
    const serialized = try stable_json.compactAlloc(arena, element);
    return std.mem.indexOf(u8, serialized, sig) != null;
}

/// Merge Ghoztty's hooks into the user's shared settings WITHOUT clobbering
/// their own hooks. Only Ghoztty's events are touched, and within each only
/// prior Ghoztty elements (matched by the signature) are replaced — every
/// other event and every non-Ghoztty element is preserved. Returns the
/// updated root value (arena semantics: input and output alias freely).
pub fn merge(
    arena: Allocator,
    base: std.json.Value,
    banner_script_path: []const u8,
) Allocator.Error!std.json.Value {
    var root = switch (base) {
        .object => |o| o,
        // The component refuses non-object files before calling; a fresh
        // object here keeps the function total for tests.
        else => std.json.ObjectMap.init(arena),
    };
    var hooks: std.json.ObjectMap = hooks: {
        if (root.get("hooks")) |existing| switch (existing) {
            .object => |o| break :hooks o,
            else => {},
        };
        break :hooks std.json.ObjectMap.init(arena);
    };

    const sig = signature(banner_script_path);
    var block = try hooksBlock(arena, banner_script_path);
    for (block.keys(), block.values()) |event, want| {
        var arr = std.json.Array.init(arena);
        if (hooks.get(event)) |existing| switch (existing) {
            .array => |a| for (a.items) |element| {
                if (!try isOurs(arena, element, sig)) try arr.append(element);
            },
            else => {},
        };
        for (want.array.items) |element| try arr.append(element);
        try hooks.put(event, .{ .array = arr });
    }

    try root.put("hooks", .{ .object = hooks });
    return .{ .object = root };
}

/// Remove ONLY Ghoztty's own elements (matched by signature) from every
/// event, pruning an event key that becomes empty and the whole `hooks` map
/// if it becomes empty. The user's other events and elements are left intact.
pub fn removeFragment(
    arena: Allocator,
    base: std.json.Value,
    banner_script_path: []const u8,
) Allocator.Error!std.json.Value {
    var root = switch (base) {
        .object => |o| o,
        else => return base,
    };
    var hooks: std.json.ObjectMap = hooks: {
        if (root.get("hooks")) |existing| switch (existing) {
            .object => |o| break :hooks o,
            else => return .{ .object = root },
        };
        return .{ .object = root };
    };

    const sig = signature(banner_script_path);
    // Snapshot the key list: removal mutates the map we would be iterating.
    const events = try arena.dupe([]const u8, hooks.keys());
    for (events) |event| {
        const value = hooks.get(event) orelse continue;
        const a = switch (value) {
            .array => |a| a,
            else => continue,
        };
        var kept = std.json.Array.init(arena);
        for (a.items) |element| {
            if (!try isOurs(arena, element, sig)) try kept.append(element);
        }
        if (kept.items.len == 0) {
            _ = hooks.orderedRemove(event);
        } else {
            try hooks.put(event, .{ .array = kept });
        }
    }

    if (hooks.count() == 0) {
        _ = root.orderedRemove("hooks");
    } else {
        try root.put("hooks", .{ .object = hooks });
    }
    return .{ .object = root };
}

/// State of ONLY Ghoztty's fragment: collect the Ghoztty elements (by
/// signature) from our events and compare them as sets against what we would
/// install. The user's other hooks vary and are never part of the compare.
pub fn fragmentState(
    arena: Allocator,
    base: std.json.Value,
    banner_script_path: []const u8,
) Allocator.Error!InstallState {
    const root = switch (base) {
        .object => |o| o,
        else => return .not_installed,
    };
    const hooks: ?std.json.ObjectMap = hooks: {
        if (root.get("hooks")) |existing| switch (existing) {
            .object => |o| break :hooks o,
            else => {},
        };
        break :hooks null;
    };

    const sig = signature(banner_script_path);
    var found = false;
    var matches = true;
    var block = try hooksBlock(arena, banner_script_path);
    for (block.keys(), block.values()) |event, want| {
        var have: std.ArrayList([]const u8) = .empty;
        if (hooks) |h| if (h.get(event)) |existing| switch (existing) {
            .array => |a| for (a.items) |element| {
                const serialized = try stable_json.compactAlloc(arena, element);
                if (std.mem.indexOf(u8, serialized, sig) != null)
                    try have.append(arena, serialized);
            },
            else => {},
        };
        if (have.items.len > 0) found = true;

        var want_serialized: std.ArrayList([]const u8) = .empty;
        for (want.array.items) |element| {
            try want_serialized.append(arena, try stable_json.compactAlloc(arena, element));
        }
        if (!setEqual(want_serialized.items, have.items)) matches = false;
    }

    if (!found) return .not_installed;
    return if (matches) .installed else .outdated;
}

/// Set equality over serialized elements (duplicates collapse, order
/// irrelevant), matching Mac's `Set` compare.
fn setEqual(a: []const []const u8, b: []const []const u8) bool {
    return subset(a, b) and subset(b, a);
}

fn subset(a: []const []const u8, b: []const []const u8) bool {
    outer: for (a) |needle| {
        for (b) |candidate| {
            if (std.mem.eql(u8, needle, candidate)) continue :outer;
        }
        return false;
    }
    return true;
}

// -----------------------------------------------------------------------------
// Tests — mirroring Mac's ClaudeHookSpecTests (minus the external-plugin
// coverage, which moves to T869 with the code it tests).
// -----------------------------------------------------------------------------

const testing = std.testing;

const test_banner = "/Users/x/.config/ghoztty/hooks/ghoztty-banner.sh";

fn parse(alloc: Allocator, text: []const u8) !std.json.Value {
    return std.json.parseFromSliceLeaky(std.json.Value, alloc, text, .{});
}

fn emptyObject(alloc: Allocator) std.json.Value {
    return .{ .object = std.json.ObjectMap.init(alloc) };
}

fn hooksOf(v: std.json.Value) ?std.json.ObjectMap {
    const root = switch (v) {
        .object => |o| o,
        else => return null,
    };
    const h = root.get("hooks") orelse return null;
    return switch (h) {
        .object => |o| o,
        else => null,
    };
}

test "merged fragment carries a 10s timeout on every command" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, emptyObject(alloc), test_banner);
    const hooks = hooksOf(merged).?;
    for (hooks.keys(), hooks.values()) |event, value| {
        _ = event;
        var commands: usize = 0;
        for (value.array.items) |element| {
            const inner = element.object.get("hooks").?.array;
            for (inner.items) |command| {
                commands += 1;
                try testing.expectEqual(@as(i64, 10), command.object.get("timeout").?.integer);
            }
        }
        try testing.expect(commands > 0);
    }
}

test "merged fragment registers every activity-state event with its verb" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, emptyObject(alloc), test_banner);
    const hooks = hooksOf(merged).?;
    const activity_path = try hook_scripts.activityStatePath(alloc, test_banner);

    const cases = [_]struct { event: []const u8, verb: []const u8 }{
        .{ .event = "SessionStart", .verb = "session-start" },
        .{ .event = "PreToolUse", .verb = "pause" },
        .{ .event = "Notification", .verb = "pause" },
        .{ .event = "PermissionRequest", .verb = "pause" },
        .{ .event = "PostToolUse", .verb = "tool-tick" },
        .{ .event = "SubagentStart", .verb = "agent-start" },
        .{ .event = "SubagentStop", .verb = "agent-stop" },
        .{ .event = "Stop", .verb = "settle" },
    };
    for (cases) |case| {
        const text = try stable_json.compactAlloc(alloc, hooks.get(case.event).?);
        try testing.expect(std.mem.indexOf(u8, text, activity_path) != null);
        const with_runtime = try std.fmt.allocPrint(alloc, "{s} --runtime=claude", .{case.verb});
        try testing.expect(std.mem.indexOf(u8, text, with_runtime) != null);
    }
}

test "PreToolUse pauses only for user-blocking tools" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, emptyObject(alloc), test_banner);
    const elements = hooksOf(merged).?.get("PreToolUse").?.array;
    try testing.expectEqual(@as(usize, 1), elements.items.len);
    try testing.expectEqualStrings(
        "AskUserQuestion|ExitPlanMode",
        elements.items[0].object.get("matcher").?.string,
    );
}

test "activity-state elements are owned by the directory signature and fully removed" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, emptyObject(alloc), test_banner);
    try testing.expectEqual(InstallState.installed, try fragmentState(alloc, merged, test_banner));

    const removed = try removeFragment(alloc, merged, test_banner);
    const text = try stable_json.compactAlloc(alloc, removed);
    try testing.expect(std.mem.indexOf(u8, text, "ghoztty-activity-state.sh") == null);
    try testing.expect(removed.object.get("hooks") == null);
}

test "a banner-only install is outdated, not installed and not absent" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const banner_only = try parse(alloc,
        \\{"hooks":{"Stop":[{"hooks":[{"type":"command",
        \\"command":"bash '/Users/x/.config/ghoztty/hooks/ghoztty-banner.sh' stop-hook --runtime=claude",
        \\"timeout":10}]}]}}
    );
    try testing.expectEqual(InstallState.outdated, try fragmentState(alloc, banner_only, test_banner));
}

test "merge preserves unrelated keys and is detectable" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, try parse(alloc,
        \\{"theme":"dark","model":"opus"}
    ), test_banner);
    try testing.expectEqualStrings("dark", merged.object.get("theme").?.string);
    try testing.expectEqualStrings("opus", merged.object.get("model").?.string);
    try testing.expectEqual(InstallState.installed, try fragmentState(alloc, merged, test_banner));
}

test "removeFragment leaves the rest and reads not_installed" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, try parse(alloc,
        \\{"theme":"dark"}
    ), test_banner);
    const removed = try removeFragment(alloc, merged, test_banner);
    try testing.expectEqualStrings("dark", removed.object.get("theme").?.string);
    try testing.expectEqual(InstallState.not_installed, try fragmentState(alloc, removed, test_banner));
}

test "merge preserves the user's own PreToolUse hooks alongside ours" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, try parse(alloc,
        \\{"theme":"dark","hooks":{"PreToolUse":[{"hooks":[{"type":"command","command":"echo mine"}]}]}}
    ), test_banner);
    const hooks = hooksOf(merged).?;
    const pretool = try stable_json.compactAlloc(alloc, hooks.get("PreToolUse").?);
    try testing.expect(std.mem.indexOf(u8, pretool, "echo mine") != null);
    try testing.expect(hooks.get("SessionStart") != null);
    try testing.expect(hooks.get("UserPromptSubmit") != null);
    try testing.expect(hooks.get("Stop") != null);
    try testing.expectEqual(InstallState.installed, try fragmentState(alloc, merged, test_banner));
}

test "removeFragment drops only Ghoztty's entries" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, try parse(alloc,
        \\{"theme":"dark","hooks":{"PreToolUse":[{"hooks":[{"type":"command","command":"echo mine"}]}]}}
    ), test_banner);
    const removed = try removeFragment(alloc, merged, test_banner);
    try testing.expectEqualStrings("dark", removed.object.get("theme").?.string);
    const hooks = hooksOf(removed).?;
    const pretool = try stable_json.compactAlloc(alloc, hooks.get("PreToolUse").?);
    try testing.expect(std.mem.indexOf(u8, pretool, "echo mine") != null);
    try testing.expect(hooks.get("SessionStart") == null);
    try testing.expect(hooks.get("UserPromptSubmit") == null);
    try testing.expect(hooks.get("Stop") == null);
    try testing.expectEqual(InstallState.not_installed, try fragmentState(alloc, removed, test_banner));
}

test "a user's own SessionStart element survives install and uninstall" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const merged = try merge(alloc, try parse(alloc,
        \\{"hooks":{"SessionStart":[{"hooks":[{"type":"command","command":"echo user-start"}]}]}}
    ), test_banner);
    const merged_ss = hooksOf(merged).?.get("SessionStart").?.array;
    // The user's one element, plus Ghoztty's two: the banner (matched to
    // `startup|clear`) and the activity-state sweep (every session start).
    try testing.expectEqual(@as(usize, 3), merged_ss.items.len);
    const merged_text = try stable_json.compactAlloc(alloc, .{ .array = merged_ss });
    try testing.expect(std.mem.indexOf(u8, merged_text, "echo user-start") != null);
    try testing.expect(std.mem.indexOf(u8, merged_text, test_banner) != null);
    try testing.expectEqual(InstallState.installed, try fragmentState(alloc, merged, test_banner));

    const removed = try removeFragment(alloc, merged, test_banner);
    const removed_ss = hooksOf(removed).?.get("SessionStart").?.array;
    try testing.expectEqual(@as(usize, 1), removed_ss.items.len);
    const removed_text = try stable_json.compactAlloc(alloc, .{ .array = removed_ss });
    try testing.expect(std.mem.indexOf(u8, removed_text, "echo user-start") != null);
    try testing.expect(std.mem.indexOf(u8, removed_text, test_banner) == null);
}

test "reinstall over a stale fragment replaces rather than duplicates" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    // An older install whose Stop event carried only the banner command.
    const stale = try parse(alloc,
        \\{"hooks":{"Stop":[{"hooks":[{"type":"command",
        \\"command":"bash '/Users/x/.config/ghoztty/hooks/ghoztty-banner.sh' stop-hook --runtime=claude",
        \\"timeout":10}]}]}}
    );
    const merged = try merge(alloc, stale, test_banner);
    // Stop holds exactly the CURRENT two elements — the stale one is gone.
    try testing.expectEqual(@as(usize, 2), hooksOf(merged).?.get("Stop").?.array.items.len);
    try testing.expectEqual(InstallState.installed, try fragmentState(alloc, merged, test_banner));
}
